clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -483,12 +505,19 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
                         parsed.add(parseRecord(tp, logEntry));
                         bytes += logEntry.size();
                     }
+
                     if (!parsed.isEmpty()) {
                         ConsumerRecord<K, V> record = parsed.get(parsed.size() - 1);
                         this.subscriptions.fetched(tp, record.offset() + 1);
-                        this.records.add(new PartitionRecords<K, V>(fetchOffset, tp, parsed));
+                        this.records.add(new PartitionRecords<>(fetchOffset, tp, parsed));
                         this.sensors.recordsFetchLag.record(partition.highWatermark - record.offset());
+                    } else if (buffer.capacity() >= this.fetchSize) {
```

Comment:
We only assure that a message is too large to be ever returned if we have already fetched the maximum size for a partition which still only contains part of that message (i.e. ```parsed.isEmpty()``` since we will not add it if it is only partial message). I think ```buffer.capacity()``` should never be > this.fetchSize but just use ```>=``` just in case.